packages/bitcoin/src/utils/helpers.ts

```diff
@@ -0,0 +1,210 @@
+import type { networks, Payment } from 'bitcoinjs-lib';
+import { payments, Psbt } from 'bitcoinjs-lib';
+
+import { TypeOfAddress } from '../types.js';
+import type {
+  BitcoinTransferInputData,
+  BitcoinTransferRequest,
+  PaymentReturnData,
+  UTXOData,
+  CreateInputData,
+  CreatePsbtParams,
+} from '../types.js';
+
+/**
+ * Get the scriptPubKey for the given public key and network
+ *
+ * @category Helpers
+ * @param typeOfAddress - type of address to use: currently p2wpkh or p2tr
+ * @param publicKey - public of the signer
+ * @param network - network to use
+ * @returns {scriptPubKey: Buffer}
+ */
+export function getScriptPubkey(
+  typeOfAddress: TypeOfAddress,
+  publicKey: Buffer,
+  network: networks.Network,
+): {
+  scriptPubKey: Buffer;
+} {
+  if (typeOfAddress === TypeOfAddress.P2WPKH) {
+    const { output } = payments.p2wpkh({
+      pubkey: publicKey,
+      network,
+    }) as PaymentReturnData;
+
+    return { scriptPubKey: output };
+  } else {
+    const { output } = payments.p2tr({
+      internalPubkey: publicKey,
+      network,
+    }) as PaymentReturnData;
+
+    return { scriptPubKey: output };
+  }
+}
+
+/**
+ * Encode the deposit address and the domain id to pass into the OP_RETURN output
+ *
+ * @category Helpers
+ * @param depositAddress - address to deposit
+ * @param destinationDomainId - destination domain id
+ * @returns {Payment}
+ */
+function encodeDepositAddress(depositAddress: string, destinationDomainId: number): Payment {
+  return payments.embed({
+    data: [Buffer.from(`${depositAddress}_${destinationDomainId}`)],
+  });
+}
+
+/**
+ * Create the input data for the PSBT
+ *
+ * @category Helpers
+ * @param utxoData - UTXO data
+ * @param publicKey - public key of the signer
+ * @param network - network to use
+ * @param typeOfAddress - type of address to use
+ * @returns {BitcoinTransferInputData}
+ */
+export function createInputData({
+  utxoData: { utxoTxId, utxoOutputIndex, utxoAmount },
+  publicKey,
+  network,
+  typeOfAddress,
+}: CreateInputData): BitcoinTransferInputData {
+  if (typeOfAddress !== TypeOfAddress.P2TR) {
+    return {
+      hash: utxoTxId as unknown as Buffer,
+      index: utxoOutputIndex,
+      witnessUtxo: {
+        script: getScriptPubkey(typeOfAddress, publicKey, network).scriptPubKey,
+        value: Number(utxoAmount),
+      },
+    };
+  }
+  return {
+    hash: utxoTxId,
+    index: utxoOutputIndex,
+    witnessUtxo: {
+      script: getScriptPubkey(typeOfAddress, publicKey, network).scriptPubKey as unknown as Buffer,
+      value: Number(utxoAmount),
+    },
+    tapInternalKey: publicKey,
+  };
+}
+
+/**
+ * Check if the amount to transfer is valid
+ * @category Helpers
+ * @param amount - amount to transfer
+ * @param utxoData - UTXO data
+ * @returns {boolean}
+ */
+const invalidAmount = (amount: bigint, utxoData: UTXOData[]): boolean => {
+  if (utxoData.length === 1) {
```

Comment:
It would be nice to have a comment, why exactly we react on the array with one data inside

Reply:
removed and we just iterate